### PR TITLE
[#7619] irodsctl stop: Fix orphan agent cleanup (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #7619

I've confirmed the fix and test validity on Ubuntu 22.04. I would like to test on Centos 7, too, because we will be releasing for that platform for 4.3.2 and it uses python2 by default.

Need to run core tests as well.

### CODE REVIEW

File: scripts/irods/test/test_irodsctl.py [hunk 1]
```diff
@@ -43,6 +42,69 @@ def test_configuration_schema_validation_from_file(self):
...
+    def test_agents_kept_alive_by_connected_clients_are_cleaned_up_after_failed_graceful_shutdown__issue_7619(self):
+        # Get the server process so that we can monitor the child processes.
+        server_proc = IrodsController().get_server_proc()
+        self.assertIsNotNone(server_proc)
+
+        with session.make_session_for_existing_admin() as admin_session:
+            user = session.mkuser_and_return_session('rodsuser', 'bilbo', 'bpass', lib.get_hostname())
+            logical_path = f'{user.session_collection}/the_replica_that_never_should_have_been'
+
+            # Use a nested try-except here because we want to ensure that the user session is cleaned up.
```
Comment (alanking) [inline scripts/irods/test/test_irodsctl.py:54]: Whoops. Remove this.

**Review by korydraughn — COMMENTED** [review]
Seems good to me.

Let's get another set of eyes on this.

**Comment by alanking** [comment]
I have confirmed that using `filter()` in python2 yields the same result as `list(filter())` in the case with which we are concerned in this issue:
```python
>>> l = [1,2,3,4]
>>> print filter(lambda x: 0 == x % 2, l)
[2, 4]
>>> print list(filter(lambda x: 0 == x % 2, l))
[2, 4]
```

File: scripts/irods/test/test_irodsctl.py (hunk 1 above)
Comment (alanking) [reply scripts/irods/test/test_irodsctl.py]: Done

**Comment by korydraughn** [comment]
Which means you don't have to worry about version sniffing, correct?

**Comment by alanking** [comment]
Correct.

**Comment by alanking** [comment]
Core tests passed.

**Comment by korydraughn** [comment]
Are you still testing centos 7?

**Comment by alanking** [comment]
I cannot build on main for Centos 7 right now, so I guess what I can do is go ahead and cherry-pick to 4-3-stable and try building there. If we still can't build then we have a problem. Will write back once I get some results.

**Comment by korydraughn** [comment]
Sounds good.

**Review by MartinFlores751 — APPROVED** [review]
Everything looks good to me.

File: scripts/irods/controller.py [hunk 2]
```diff
@@ -293,7 +293,7 @@ def should_return_proc(_p):
 
     if server_proc.is_running():
         try:
-            cur_descendants = filter(should_return_proc, server_proc.children(recursive=True))
+            cur_descendants = list(filter(should_return_proc, server_proc.children(recursive=True)))
```
Comment (FifthPotato) [inline scripts/irods/controller.py:296]: If we're using `difference` below, does that mean `server_descendants` is a `set`? Would it make more sense to use a `set` here to compute `set` difference?

**Review by FifthPotato — COMMENTED** [review]
Looks pretty good to me. One possible nitpick is all I got!

File: scripts/irods/controller.py (hunk 2 above)
Comment (alanking) [reply scripts/irods/controller.py:296]: `server_descendants` is indeed a `set`. I'll look into usin `set` here instead.

Comment (alanking) [reply scripts/irods/controller.py:296]: Yep, that works. Good catch

**Comment by alanking** [comment]
I have confirmed that the existing fix works on Centos 7 (which also means that 4-3-stable builds for Centos 7, hooray). I also confirmed that the `set` instead of `list` works.

**Comment by korydraughn** [comment]
Seems ready to go. Is there anything left to do for this?

**Comment by alanking** [comment]
I don't think so. I'll go ahead and squash with no #s.

**Review by alanking — COMMENTED** [review]
Hmm... wait a second, something weird happened. Hold please...

File: scripts/irods/controller.py [hunk 3]
```diff
@@ -293,7 +293,7 @@ def should_return_proc(_p):
 
     if server_proc.is_running():
         try:
-            cur_descendants = filter(should_return_proc, server_proc.children(recursive=True))
+            cur_descendants = set(filter(should_return_proc, server_proc.children(recursive=True)))
```
Comment (alanking) [inline scripts/irods/controller.py:296]: This is t

Comment (alanking) [reply scripts/irods/controller.py:296]: Fixed.

**Comment by alanking** [comment]
Okay, fixed the weird thing and confirmed that the test still passes.

**Review by korydraughn — APPROVED** [review]
Seems ready to pound.

Go for it.

**Comment by alanking** [comment]
#'d, and mergin.

**Pull request merged**